receiver/apachedruidreceiver/metrics.go

```diff
@@ -0,0 +1,308 @@
+	config         *MetricsConfig
+	server         *http.Server
+	consumer       consumer.Metrics
+	wg             *sync.WaitGroup
+	id             component.ID // ID of the receiver component
+	metricsBuilder *metadata.MetricsBuilder
+}
+
+func newMetricsReceiver(params receiver.CreateSettings, cfg *Config, consumer consumer.Metrics) (*metricsReceiver, error) {
+	var tlsConfig *tls.Config
+	recv := &metricsReceiver{
+		config:         &cfg.Metrics,
+		consumer:       consumer,
+		logger:         params.Logger,
+		wg:             &sync.WaitGroup{},
+		id:             params.ID,
+		metricsBuilder: metadata.NewMetricsBuilder(cfg.Metrics.MetricsBuilderConfig, params),
+	}
+
+	if recv.config.TLS != nil {
+		var err error
+
+		tlsConfig, err = recv.config.TLS.LoadTLSConfig()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	s := &http.Server{
+		TLSConfig:         tlsConfig,
+		Handler:           http.HandlerFunc(recv.handleRequest),
+		ReadHeaderTimeout: 20 * time.Second,
+	}
+
+	recv.server = s
+	return recv, nil
+}
+
+func (mReceiver *metricsReceiver) Start(ctx context.Context, host component.Host) error {
+	return mReceiver.startListening(ctx, host)
+}
+
+func (mReceiver *metricsReceiver) Shutdown(ctx context.Context) error {
+	mReceiver.logger.Debug("Shutting down server")
+	err := mReceiver.server.Shutdown(ctx)
+	if err != nil {
+		return err
+	}
+
+	mReceiver.logger.Debug("Waiting for shutdown to complete.")
+	mReceiver.wg.Wait()
+	return nil
+}
+
+func (mReceiver *metricsReceiver) startListening(ctx context.Context, host component.Host) error {
+	mReceiver.logger.Debug("starting receiver HTTP server")
+	// We use l.server.Serve* over l.server.ListenAndServe*
+	// So that we can catch and return errors relating to binding to network interface on start.
+	var listenConfig net.ListenConfig
+
+	listener, err := listenConfig.Listen(ctx, "tcp", mReceiver.config.Endpoint)
+	if err != nil {
+		return err
+	}
+
+	mReceiver.wg.Add(1)
+	if mReceiver.config.TLS != nil {
+		go func() {
+			defer mReceiver.wg.Done()
+
+			mReceiver.logger.Debug("Starting ServeTLS",
+				zap.String("address", mReceiver.config.Endpoint),
+				zap.String("certfile", mReceiver.config.TLS.CertFile),
+				zap.String("keyfile", mReceiver.config.TLS.KeyFile))
+
+			err := mReceiver.server.ServeTLS(listener, mReceiver.config.TLS.CertFile, mReceiver.config.TLS.KeyFile)
+
+			mReceiver.logger.Debug("Serve TLS done")
+
+			if err != http.ErrServerClosed {
+				mReceiver.logger.Error("ServeTLS failed", zap.Error(err))
+				host.ReportFatalError(err)
+			}
+		}()
+	} else {
+		go func() {
+			defer mReceiver.wg.Done()
+
+			mReceiver.logger.Debug("Starting Serve",
+				zap.String("address", mReceiver.config.Endpoint))
+
+			err = mReceiver.server.Serve(listener)
+
+			mReceiver.logger.Debug("Serve done")
+
+			if err != http.ErrServerClosed {
+				mReceiver.logger.Error("Serve failed", zap.Error(err))
+				host.ReportFatalError(err)
+			}
+		}()
+	}
+
+	return nil
```

Comment:
```suggestion
	if mReceiver.config.TLS != nil {
		go func() {
			defer mReceiver.wg.Done()

			mReceiver.logger.Debug("Starting ServeTLS",
				zap.String("address", mReceiver.config.Endpoint),
				zap.String("certfile", mReceiver.config.TLS.CertFile),
				zap.String("keyfile", mReceiver.config.TLS.KeyFile))

			err := mReceiver.server.ServeTLS(listener, mReceiver.config.TLS.CertFile, mReceiver.config.TLS.KeyFile)

			mReceiver.logger.Debug("Serve TLS done")

			if err != http.ErrServerClosed {
				mReceiver.logger.Error("ServeTLS failed", zap.Error(err))
				host.ReportFatalError(err)
			}
		}()
		return nil
	} 
			go func() {
			defer mReceiver.wg.Done()

			mReceiver.logger.Debug("Starting Serve",
				zap.String("address", mReceiver.config.Endpoint))

			err = mReceiver.server.Serve(listener)

			mReceiver.logger.Debug("Serve done")

			if err != http.ErrServerClosed {
				mReceiver.logger.Error("Serve failed", zap.Error(err))
				host.ReportFatalError(err)
			}
		}()
	

	return nil
```

Its pretty idiomatic to return early if there is nothing else to do for that conditional: https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow